tensorflow_probability/python/mcmc/replica_exchange_mc.py

```diff
@@ -0,0 +1,426 @@
+                            num_replica=inverse_temperatures.shape[0],
+                            exchange_proposed_fn=exchange_proposed_fn,
+                            seed=seed, name=name)
+    self.replica_kernels = []
+    for i in range(self.num_replica):
+      self._seed_stream = distributions_util.gen_new_seed(
+          self._seed_stream, salt='replica_kernels')
+      self.replica_kernels.append(replica_kernel_class(
+          target_log_prob_fn=_replica_log_prob_fn(
+              inverse_temperatures[i], target_log_prob_fn),
+          seed=self._seed_stream, name=name, **kwargs))
+
+  @property
+  def target_log_prob_fn(self):
+    return self._parameters['target_log_prob_fn']
+
+  @property
+  def inverse_temperatures(self):
+    return self._parameters['inverse_temperatures']
+
+  @property
+  def num_replica(self):
+    return self._parameters['num_replica']
+
+  @property
+  def exchange_proposed_fn(self):
+    return self._parameters['exchange_proposed_fn']
+
+  @property
+  def seed(self):
+    return self._parameters['seed']
+
+  @property
+  def name(self):
+    return self._parameters['name']
+
+  @property
+  def parameters(self):
+    """Return `dict` of ``__init__`` arguments and their values."""
+    return self._parameters
+
+  @property
+  def is_calibrated(self):
+    return True
+
+  def one_step(self, current_state, previous_kernel_results):
+    """Takes one step of the TransitionKernel.
+
+    Args:
+      current_state: `Tensor` or Python `list` of `Tensor`s representing the
+        current state(s) of the Markov chain(s).
+      previous_kernel_results: A (possibly nested) `tuple`, `namedtuple` or
+        `list` of `Tensor`s representing internal calculations made within the
+        previous call to this function (or as returned by `bootstrap_results`).
+
+    Returns:
+      next_state: `Tensor` or Python `list` of `Tensor`s representing the
+        next state(s) of the Markov chain(s).
+      kernel_results: A (possibly nested) `tuple`, `namedtuple` or `list` of
+        `Tensor`s representing internal calculations made within this function.
+        This inculdes replica states.
+    """
+    with tf.name_scope(
+        name=mcmc_util.make_name(self.name, 'remc', 'one_step'),
+        values=[current_state, previous_kernel_results]):
+      sampled_replica_states = []
+      sampled_replica_results = []
+      sampled_replica_ratios = []
+      for i in range(self.num_replica):
+        [
+            sampled_state,
+            sampled_results,
+        ] = self.replica_kernels[i].one_step(
+            previous_kernel_results.replica_states[i],
+            previous_kernel_results.replica_results[i])
+        sampled_replica_states.append(sampled_state)
+        sampled_replica_results.append(sampled_results)
+        if not mcmc_util.is_list_like(sampled_state):
+          sampled_state = [sampled_state]
+        sampled_replica_ratios.append(self.target_log_prob_fn(*sampled_state))
```

Comment:
Thanks for advice!

I am trying to implement, but the result does not go well.

To recalculate ratios, the code and sampling result is this.
```Python
sampled_replica_ratios = [self.target_log_prob_fn(*srs)
                          if mcmc_util.is_list_like(srs)
                          else self.target_log_prob_fn(*[srs])
                          for i, srs in enumerate(sampled_replica_states)]
sampled_replica_ratios = tf.stack(sampled_replica_ratios, axis=0)
```
![image](https://user-images.githubusercontent.com/17809477/38678580-652c72fa-3e9c-11e8-9f79-203fdbc2e998.png)


But to avoid recalculation, the code and sampling result is this.
```Python
sampled_replica_ratios = \
    [srr.target_log_prob if 'target_log_prob' in srr._fields
     else srr.accepted_results.target_log_prob
     for i, srr in enumerate(sampled_replica_results)]
sampled_replica_ratios = \
    tf.stack(sampled_replica_ratios, axis=0) / self.inverse_temperatures
```
![image](https://user-images.githubusercontent.com/17809477/38678587-6b292a86-3e9c-11e8-9c88-f1f31938d2b7.png)

I think it happens because the error comes out from the original value when dividing after multiplication.

I think that it would be better to recalculate, for accuracy.
How about this?

Reply:
The two should be identical so this suggests something strange is going on.

One thing that stands out to me. I _think_ you want:
     tf.stack(sampled_replica_ratios, axis=-1) / self.inverse_temperatures
Broadcasting pads 1's to the left of a shape, not the right. By stacking on the rightmost dim, this ensures that the inverse_temps are broadcast across chains.

Having said that, I doubt this will fix the problem because I think it would have errored out if this was really the problem. (Although if you happened to have the same number of chains as replicas then this could indeed be the culprit.)

Since I dont have any other ideas if youre unable to track down the bug, I will look into it post submission and see if I can find whats going on.  This is worth fixing because we must assume target_log_prob_fn is expensive to call.  (However, that doesnt mean we need to make this optimization on our first attempt.)

Reply:
Thanks for the comments!

I tracked down and fixed the bug.
Although I must exchange inverse_temp of `next_replica_results[i].target_log_prob` of `next_replica_results[i].accepted_results.target_log_prob` after exchange replica results and states, I didn't.
I fixed not to recalculate.

Reply:
Great!  This is a big savings, so Im glad you were able to identify (and correct!) the problem.